problems with setting up mbdetls as transport layer

We have tried to port the demo application but the connection is nor working and we get an error for invalid handler due to invalid certificates:
mqttClientAws.c connectToServerWithBackoffRetries() 640 | rootCA /AmazonRootCA1.pem
mqttClientAws.c connectToServerWithBackoffRetries() 641 | clientCRT  /xxxxxxx-certificate.pem.crt
mqttClientAws.c connectToServerWithBackoffRetries() 642 | clientKey  /xxxxxxx-private.pem.crt
mqttClientAws.c connectToServerWithBackoffRetries() 654 | ===SUCCESSFULLY INITIALIZED PKCS #11.===
mqttClientAws[5879] | [DEBUG] [PKCS11] [core_pkcs11_mbedtls.c:483] Successfully found object class attribute.^M
mqttClientAws[5879] | [INFO] [PKCS11] [core_pkcs11_mbedtls.c:2823] Creating a 0x3 type object.^M
mqttClientAws[5879] | [DEBUG] [PKCS11] [core_pkcs11_mbedtls.c:2057] Successfully found the key type in the template.^M
mqttClientAws[5879] | [DEBUG] [PKCS11] [core_pkcs11_mbedtls.c:2086] Successfully found the label in the template.^M
mqttClientAws[5879] | [DEBUG] [PKCS11] [core_pkcs11_mbedtls.c:1259] Key was private type.^M
mqttClientAws[5879] | [DEBUG] [PKCS11] [core_pkcs11_mbedtls.c:1268] Received RSA key type.^M
mqttClientAws[5879] | [DEBUG] [PKCS11] [core_pkcs11_mbedtls.c:1288] Allocating a 1200 bytes sized buffer to write the key to.^M
mqttClientAws[5879] | [DEBUG] [PKCS11] [core_pkcs11_pal_utils.c:123] Converted Device Priv TLS Key to corePKCS11_Key.dat^M
mqttClientAws[5879] | Successfully wrote 1192 to corePKCS11_Key.dat[DEBUG] [PKCS11] [core_pkcs11_mbedtls.c:3352] Search parameters other than label are ignored.^M
mqttClientAws[5879] | [DEBUG] [PKCS11] [core_pkcs11_mbedtls.c:3446] Could not find the object handle in the list. Trying to search PKCS #11 PAL for object.^M
mqttClientAws[5879] | [DEBUG] [PKCS11] [core_pkcs11_pal_utils.c:123] Converted Device Cert to corePKCS11_Certificate.dat^M
mqttClientAws[5879] | [INFO] [PKCS11] [core_pkcs11_pal.c:63] Could not open corePKCS11_Certificate.dat for reading.^M
mqttClientAws[5879] | [ERROR] [PKCS11] [core_pkcs11.c:370] xFindObjectWithLabelAndClass ERROR CK_INVALID_HANDLE^M
mqttClientAws[5879] | [INFO] [FLEET_PROVISIONING_DEMO] [pkcs11_operations.c:770] Writing certificate into label "Device Cert".^M
mqttClientAws[5879] | [DEBUG] [PKCS11] [core_pkcs11_mbedtls.c:483] Successfully found object class attribute.^M
mqttClientAws[5879] | [INFO] [PKCS11] [core_pkcs11_mbedtls.c:2823] Creating a 0x1 type object.^M
mqttClientAws[5879] | [DEBUG] [PKCS11] [core_pkcs11_pal_utils.c:123] Converted Device Cert to corePKCS11_Certificate.dat^M
mqttClientAws[5879]/mqttClientAwsComponent T=main | mqttClientAws.c connectToServerWithBackoffRetries() 668 | ===SUCCESSFULLY PROVISIONED PKCS #11.===
mqttClientAws.c connectToServerWithBackoffRetries() 686 | Establishing a TLS session to axxxxxx-ats.iot.ap-southeast-2.amazonaws.com:443.
mqttClientAws[5879] | Successfully wrote 861 to corePKCS11_Certificate.dat[DEBUG] [PKCS11] [core_pkcs11_mbedtls.c:3352] Search parameters other than label are ignored.^M
mqttClientAws[5879] | [DEBUG] [PKCS11] [core_pkcs11_mbedtls.c:3352] Search parameters other than label are ignored.^M
mqttClientAws[5879] | [ERROR] [PKCS11] [core_pkcs11.c:370] xFindObjectWithLabelAndClass ERROR CK_INVALID_HANDLE^M
mqttClientAws[5879] | [ERROR] [Transport_MbedTLS_PKCS11] [mbedtls_pkcs11_posix.c:625] Function returned ERROR.^M
mqttClientAws[5879] | [ERROR] [Transport_MbedTLS_PKCS11] [mbedtls_pkcs11_posix.c:400] Failed to setup key handling by PKCS #11.^M
mqttClientAws[5879] | [WARN] [DEMO] [mqttClientAws.c:724] Connection to the broker failed. Retrying connection after 270 ms backoff.^M

This is the code we are using, which we got partly from the mqtt_operation.c in the repository:
BackoffAlgorithmStatus_t backoffAlgStatus = BackoffAlgorithmSuccess;
  MbedtlsPkcs11Status_t tlsStatus = MBEDTLS_PKCS11_SUCCESS;
  BackoffAlgorithmContext_t reconnectParams;
  MbedtlsPkcs11Credentials_t tlsCredentials = { 0 };
  const char * alpn[] = { ALPN_PROTOCOL_NAME, NULL };

  /* Set the pParams member of the network context with desired transport. */
  pNetworkContext->pParams = &tlsContext;

  /* Initialize credentials for establishing TLS session. */
  tlsCredentials.pRootCaPath = rootCA;
  tlsCredentials.pClientCertLabel = clientCRT;
  tlsCredentials.pPrivateKeyLabel = clientKey;
  pkcs11ret = xInitializePkcs11Session( &tlsCredentials.p11Session );
  if( pkcs11ret != CKR_OK )
  {
      LogError( ( "Failed to initialize PKCS #11." ) );
  }

  /* AWS IoT requires devices to send the Server Name Indication (SNI)
   * extension to the Transport Layer Security (TLS) protocol and provide
   * the complete endpoint address in the host_name field. Details about
   * SNI for AWS IoT can be found in the link below.
   * https://docs.aws.amazon.com/iot/latest/developerguide/transport-security.html
   */
  tlsCredentials.disableSni = false;
  uint16_t port = AWS_IOT_MQTT_PORT;

  if( port == 443 )
  {
      /* Pass the ALPN protocol name depending on the port being used.
       * Please see more details about the ALPN protocol for AWS IoT MQTT endpoint
       * in the link below.
       * https://aws.amazon.com/blogs/iot/mqtt-with-tls-client-authentication-on-port-443-why-it-is-useful-and-how-it-works/
       */
      tlsCredentials.pAlpnProtos = alpn;
  }
bool status = false;
    if( pkcs11ret != CKR_OK )
    {
        printf( "Failed to initialize PKCS #11.");
        status = false;
    }
    else
    {
        printf( "===SUCCESSFULLY INITIALIZED PKCS #11.===");
        // Insert the claim credentials into the PKCS #11 module 
        status = loadClaimCredentials( tlsCredentials.p11Session,
                                             clientCRT,
                                           pkcs11configLABEL_DEVICE_CERTIFICATE_FOR_TLS,
                                           clientKey,
                                           pkcs11configLABEL_DEVICE_PRIVATE_KEY_FOR_TLS );


        if( status == false )
        {
            printf( "Failed to provision PKCS #11 with claim credentials." );
        }
        else 
            printf( "===SUCCESSFULLY PROVISIONED PKCS #11.===");

    }
/* Initialize reconnect attempts and interval */
BackoffAlgorithm_InitializeParams( &reconnectParams,
                                   CONNECTION_RETRY_BACKOFF_BASE_MS,
                                   CONNECTION_RETRY_MAX_BACKOFF_DELAY_MS,
                                   CONNECTION_RETRY_MAX_ATTEMPTS );

do
{
    /* Establish a TLS session with the MQTT broker. This example connects
     * to the MQTT broker as specified in BROKER_ENDPOINT and BROKER_PORT at
     * the top of this file. */
    printf("Establishing a TLS session to %.*s:%d.",
               BROKER_ENDPOINT_LENGTH,
               BROKER_ENDPOINT,
               AWS_IOT_MQTT_PORT );
    tlsStatus = Mbedtls_Pkcs11_Connect( pNetworkContext,
                                BROKER_ENDPOINT,
                                port,
                                &tlsCredentials,
                                TRANSPORT_SEND_RECV_TIMEOUT_MS );

  if( ( NULL != pxHandle ) && ( ulCount == 0UL ) )
      {
          *pxHandle = CK_INVALID_HANDLE;
          LogError(("xFindObjectWithLabelAndClass ERROR CK_INVALID_HANDLE"));
      }

Which we don`t understand why, same certificates we are using with the older SDK for embedded C from aws.
Any help will be very much appreciated!!!

we have solved the problem with initialising the the mbedtls with using a different template based on mbedtls, so we have less things to port.
It seem that we can connect to the backend with the Mqtt_init function but when we try with the Mqtt_Connect this will fail after the timeout period with "MQTTNoDataAvailable".


rootCA /home/root//home/root/AmazonRootCA1.pem
clientCRT /home/root/certificate.pem.crt
clientKey /home/root/private.pem.key
mqttClientAws[2427] | [DEBUG] [PKCS11] [core_pkcs11_mbedtls.c:461] PKCS #11 module was successfully initialized.^M
mqttClientAws[2427] | [INFO] [PKCS11] [core_pkcs11_mbedtls.c:1403] PKCS #11 successfully initialized.^M
mqttClientAws[2427] | [DEBUG] [PKCS11] [core_pkcs11_mbedtls.c:1617] Successfully Returned a PKCS #11 slot with ID 1 with a count of 1.^M
mqttClientAws[2427] | [DEBUG] [PKCS11] [core_pkcs11_mbedtls.c:1852] Assigned a 0x2 Type Session.^M
mqttClientAws[2427] | [DEBUG] [PKCS11] [core_pkcs11_mbedtls.c:1863] Assigned Mechanisms to no operation in progress.^M
mqttClientAws[2427] | [DEBUG] [PKCS11] [core_pkcs11_mbedtls.c:1879] Current session count at 0^M
mqttClientAws[2427] | [DEBUG] [PKCS11] [core_pkcs11_mbedtls.c:1965] C_Login is not implemented.^M
mqttClientAws[2427] | [DEBUG] [PKCS11] [pkcs11_operations.c:808] loadClaimCredentials readFile claimCert, 1^M
mqttClientAws[2427] | [DEBUG] [PKCS11] [pkcs11_operations.c:814] loadClaimCredentials readFile claimPrivateKey, 1^M
mqttClientAws[2427] | [DEBUG] [PKCS11] [core_pkcs11_mbedtls.c:483] Successfully found object class attribute.^M
mqttClientAws[2427] | [INFO] [PKCS11] [core_pkcs11_mbedtls.c:2823] Creating a 0x3 type object.^M
mqttClientAws[2427] | [DEBUG] [PKCS11] [core_pkcs11_mbedtls.c:2057] Successfully found the key type in the template.^M
mqttClientAws[2427] | [DEBUG] [PKCS11] [core_pkcs11_mbedtls.c:2086] Successfully found the label in the template.^M
mqttClientAws[2427] | [DEBUG] [PKCS11] [core_pkcs11_mbedtls.c:1259] Key was private type.^M
mqttClientAws[2427] | [DEBUG] [PKCS11] [core_pkcs11_mbedtls.c:1268] Received RSA key type.^M
mqttClientAws[2427] | [DEBUG] [PKCS11] [core_pkcs11_mbedtls.c:1288] Allocating a 1200 bytes sized buffer to write the key to.^M
mqttClientAws[2427] | [DEBUG] [PKCS11] [core_pkcs11_pal_utils.c:123] Converted Device Priv TLS Key to corePKCS11_Key.dat^M
mqttClientAws[2427] | Successfully wrote 1192 to corePKCS11_Key.dat[DEBUG] [PKCS11] [pkcs11_operations.c:823] loadClaimCredentials provisionPrivateKey, 1^M
mqttClientAws[2427] | [DEBUG] [PKCS11] [core_pkcs11_mbedtls.c:3352] Search parameters other than label are ignored.^M
mqttClientAws[2427] | [DEBUG] [PKCS11] [core_pkcs11_mbedtls.c:3446] Could not find the object handle in the list. Trying to search PKCS #11 PAL for object.^M
mqttClientAws[2427] | [DEBUG] [PKCS11] [core_pkcs11_pal_utils.c:123] Converted Device Cert to corePKCS11_Certificate.dat^M
mqttClientAws[2427] | [INFO] [PKCS11] [core_pkcs11_pal.c:63] Could not open corePKCS11_Certificate.dat for reading.^M
mqttClientAws[2427] | [INFO] [PKCS11] [pkcs11_operations.c:770] Writing certificate into label &quot;Device Cert&quot;.^M
mqttClientAws[2427] | [DEBUG] [PKCS11] [core_pkcs11_mbedtls.c:483] Successfully found object class attribute.^M
mqttClientAws[2427] | [INFO] [PKCS11] [core_pkcs11_mbedtls.c:2823] Creating a 0x1 type object.^M
mqttClientAws[2427] | [DEBUG] [PKCS11] [core_pkcs11_pal_utils.c:123] Converted Device Cert to corePKCS11_Certificate.dat^M
mqttClientAws[2427] | Successfully wrote 861 to corePKCS11_Certificate.dat[DEBUG] [PKCS11] [pkcs11_operations.c:832] loadClaimCredentials provisionCertificate, 1^M
mqttClientAws[2427] | [INFO] [DEMO] [mqttClientAws.c:1217] Establishing MQTT session with claim certificate...^M
mqttClientAws[2427] | [DEBUG] [MQTT] [mqtt_operations.c:376] Establishing a TLS session to a1ui4kt3m4uic8-ats.iot.ap-southeast-2.amazonaws.com:443.^M
mqttClientAws[2427] | [DEBUG] [PKCS11] [core_pkcs11_mbedtls.c:3352] Search parameters other than label are ignored.^M
mqttClientAws[2427] | [DEBUG] [PKCS11] [core_pkcs11_mbedtls.c:1053] Found object in object list matching label.^M
mqttClientAws[2427] | [DEBUG] [PKCS11] [core_pkcs11_mbedtls.c:3452] Found object in PAL. Adding object handle to list.^M
mqttClientAws[2427] | [DEBUG] [PKCS11] [core_pkcs11_mbedtls.c:1085] Found object in list by handle.^M
mqttClientAws[2427] | [DEBUG] [PKCS11] [core_pkcs11_mbedtls.c:3018] Object was a private key.^M
mqttClientAws[2427] | [DEBUG] [PKCS11] [core_pkcs11_mbedtls.c:3352] Search parameters other than label are ignored.^M
mqttClientAws[2427] | [DEBUG] [PKCS11] [core_pkcs11_mbedtls.c:1053] Found object in object list matching label.^M
mqttClientAws[2427] | [DEBUG] [PKCS11] [core_pkcs11_mbedtls.c:3452] Found object in PAL. Adding object handle to list.^M
mqttClientAws[2427] | [DEBUG] [PKCS11] [core_pkcs11_mbedtls.c:1085] Found object in list by handle.^M
mqttClientAws[2427] | [DEBUG] [PKCS11] [core_pkcs11_mbedtls.c:3034] Object was a certificate.^M
mqttClientAws[2427] | [DEBUG] [PKCS11] [core_pkcs11_mbedtls.c:1085] Found object in list by handle.^M
mqttClientAws[2427] | [DEBUG] [PKCS11] [core_pkcs11_mbedtls.c:3034] Object was a certificate.^M
mqttClientAws[2427] | [DEBUG] [Transport_MbedTLS_PKCS11] [mbedtls_pkcs11_posix.c:355] Configured MbedTLS context.^M
mqttClientAws[2427] | [DEBUG] [PKCS11] [core_pkcs11_mbedtls.c:5622] Successfully generated 28 random bytes.^M
mqttClientAws[2427] | [DEBUG] [PKCS11] [core_pkcs11_mbedtls.c:5622] Successfully generated 32 random bytes.^M
mqttClientAws[2427] | [DEBUG] [PKCS11] [core_pkcs11_mbedtls.c:5622] Successfully generated 32 random bytes.^M
mqttClientAws[2427] | [DEBUG] [PKCS11] [core_pkcs11_mbedtls.c:5622] Successfully generated 32 random bytes.^M
mqttClientAws[2427] | [DEBUG] [PKCS11] [core_pkcs11_mbedtls.c:5622] Successfully generated 32 random bytes.^M
mqttClientAws[2427] | [DEBUG] [PKCS11] [core_pkcs11_mbedtls.c:5622] Successfully generated 32 random bytes.^M
mqttClientAws[2427] | [INFO] [Transport_MbedTLS_PKCS11] [mbedtls_pkcs11_posix.c:857] TLS Connection to axxxxxxxx-ats.iot.ap-southeast-2.amazonaws.com established.^M
mqttClientAws[1332] | [DEBUG] [MQTT] [mqtt_operations.c:651] Connected to MQTT broker a1ui4kt3m4uic8-ats.iot.ap-southeast-2.amazonaws.com.^M
mqttClientAws[1332] | [DEBUG] [MQTT] [mqtt_operations.c:682] MQTT init status success!!^M
mqttClientAws[1332] | [DEBUG] [MQTT] [core_mqtt_serializer.c:459] Encoded size for length 33 is 1 bytes.^M
mqttClientAws[1332] | [DEBUG] [MQTT] [core_mqtt_serializer.c:1600] CONNECT packet remaining length=33 and packet size=35.^M
mqttClientAws[1332] | [DEBUG] [MQTT] [core_mqtt.c:1784] CONNECT packet size is 35 and remaining length is 33.^M
mqttClientAws[1332] | [DEBUG] [MQTT] [core_mqtt_serializer.c:459] Encoded size for length 33 is 1 bytes.^M
mqttClientAws[1332] | [DEBUG] [MQTT] [core_mqtt_serializer.c:1499] Length of serialized CONNECT packet is 35.^M
mqttClientAws[1332] | [DEBUG] [MQTT] [core_mqtt.c:645] BytesSent=35, BytesRemaining=0^M
mqttClientAws[1332] | [DEBUG] [MQTT] [core_mqtt.c:666] Successfully sent packet at time 95250.^M
mqttClientAws[1332] | [DEBUG] [MQTT] [core_mqtt.c:1809] Sent 35 bytes of CONNECT packet.^M
mqttClientAws[1332] | [DEBUG] [Transport_MbedTLS_PKCS11] [mbedtls_pkcs11_posix.c:924] Failed to read data. However, a read can be retried on this error. mbedTLSError= SSL - The operation timed out : <No-Low-Level-Code>.
mqttClientAws[1332] | [ERROR] [MQTT] [core_mqtt.c:1570] CONNACK recv failed with status = MQTTNoDataAvailable.^M
mqttClientAws[1332] | [ERROR] [MQTT] [core_mqtt.c:1841] MQTT connection failed with status = MQTTNoDataAvailable.^M
mqttClientAws[1332] | [ERROR] [MQTT] [mqtt_operations.c:732] Connection with MQTT broker failed with status MQTTNoDataAvailable.^M
mqttClientAws[1332] | [ERROR] [DEMO] [mqttClientAws.c:1225] Failed to establish MQTT session.^M

So is not entirely clear this is due to an incorrect setup or the backend is not answering to the CONNECT packet?

Hello sorry for the delay,
which demo setup were you following ? 

We tried this one:  https://aws.github.io/aws-iot-device-sdk-embedded-C/202108.00/docs/doxygen/output/html/fleet_provisioning_demo.html 

and we had a different error
```
"[ERROR] [Transport_MbedTLS_PKCS11] [mbedtls_pkcs11_posix.c:915] Failed to read data: mbedTLSError= SSL - The peer notified us that the connection is going to be closed : <No-Low-Level-Code>."
```
We will try with a different setup and update the case.

we have increased the log value for the logging of mbetls, this is what we see:
Jul 21 07:03:39 | mqttClientAws[7705] | [INFO] [Transport_MbedTLS_PKCS11] [mbedtls_pkcs11_posix.c:857] TLS Connection to a1ui4kt3m4uic8-ats.iot.ap-southeast-2.amazonaws.com established.^M
Jul 21 07:03:39  | mqttClientAws[7705] | [DEBUG] [MQTT] [mqtt_operations.c:651] Connected to MQTT broker a1ui4kt3m4uic8-ats.iot.ap-southeast-2.amazonaws.com.^M
Jul 21 07:03:39  | mqttClientAws[7705] | [DEBUG] [MQTT] [mqtt_operations.c:665] current time is :2903390^M
Jul 21 07:03:39  | mqttClientAws[7705] | [DEBUG] [MQTT] [mqtt_operations.c:682] MQTT init status success!!^M
Jul 21 07:03:39  | mqttClientAws[7705] | [DEBUG] [MQTT] [core_mqtt_serializer.c:459] Encoded size for length 17 is 1 bytes.^M
Jul 21 07:03:39  | mqttClientAws[7705] | [DEBUG] [MQTT] [core_mqtt_serializer.c:1600] CONNECT packet remaining length=17 and packet size=19.^M
Jul 21 07:03:39  | mqttClientAws[7705] | [DEBUG] [MQTT] [core_mqtt.c:1784] CONNECT packet size is 19 and remaining length is 17.^M
Jul 21 07:03:39  | mqttClientAws[7705] | [DEBUG] [MQTT] [core_mqtt_serializer.c:459] Encoded size for length 17 is 1 bytes.^M
Jul 21 07:03:39  | mqttClientAws[7705] | [DEBUG] [MQTT] [core_mqtt_serializer.c:1499] Length of serialized CONNECT packet is 19.^M
Jul 21 07:03:39  | mqttClientAws[7705] | mbedTLS: |2| => write
Jul 21 07:03:39  | mqttClientAws[7705] | mbedTLS: |2| => write record
Jul 21 07:03:39  | mqttClientAws[7705] | mbedTLS: |2| => encrypt buf
Jul 21 07:03:39  | mqttClientAws[7705] | mbedTLS: |4| dumping 'before encrypt: output payload' (19 bytes)
Jul 21 07:03:39  | mqttClientAws[7705] | mbedTLS: |4| 0000:  10 11 00 04 4d 51 54 54 04 00 00 3c 00 05 46 58  ....MQTT...<..FX
Jul 21 07:03:39  | mqttClientAws[7705] | mbedTLS: |4| 0010:  33 30 53                                         30S
Jul 21 07:03:39  | mqttClientAws[7705] | mbedTLS: |4| dumping 'IV used (internal)' (12 bytes)
Jul 21 07:03:39  | mqttClientAws[7705] | mbedTLS: |4| 0000:  a0 38 f6 92 00 00 00 00 00 00 00 01              .8..........
Jul 21 07:03:39  | mqttClientAws[7705] | mbedTLS: |4| dumping 'IV used (transmitted)' (8 bytes)
Jul 21 07:03:39  | mqttClientAws[7705] | mbedTLS: |4| 0000:  00 00 00 00 00 00 00 01                          ........
Jul 21 07:03:39  | mqttClientAws[7705] | mbedTLS: |4| dumping 'additional data used for AEAD' (13 bytes)
Jul 21 07:03:39  | mqttClientAws[7705] | mbedTLS: |4| 0000:  00 00 00 00 00 00 00 01 17 03 03 00 13           .............
Jul 21 07:03:39  | mqttClientAws[7705] | mbedTLS: |3| before encrypt: msglen = 19, including 0 bytes of padding
Jul 21 07:03:39  | mqttClientAws[7705] | mbedTLS: |4| dumping 'after encrypt: tag' (16 bytes)
Jul 21 07:03:39  | mqttClientAws[7705] | mbedTLS: |4| 0000:  e7 f9 a2 79 3d 42 00 0f 58 c1 7c cd 91 ed c2 60  ...y=B..X.|....`
Jul 21 07:03:39  | mqttClientAws[7705] | mbedTLS: |2| <= encrypt buf
Jul 21 07:03:39  | mqttClientAws[7705] | mbedTLS: |3| output record: msgtype = 23, version = [3:3], msglen = 43
Jul 21 07:03:39  | mqttClientAws[7705] | mbedTLS: |4| dumping 'output record sent to network' (48 bytes)
Jul 21 07:03:39  | mqttClientAws[7705] | mbedTLS: |4| 0000:  17 03 03 00 2b 00 00 00 00 00 00 00 01 7d 68 9a  ....+........}h.
Jul 21 07:03:39  | mqttClientAws[7705] | mbedTLS: |4| 0010:  9f cb 38 89 7c af fb 22 20 97 aa 24 fb 36 16 22  ..8.|.." ..$.6."
Jul 21 07:03:39  | mqttClientAws[7705] | mbedTLS: |4| 0020:  e7 f9 a2 79 3d 42 00 0f 58 c1 7c cd 91 ed c2 60  ...y=B..X.|....`
Jul 21 07:03:39  | mqttClientAws[7705] | mbedTLS: |2| => flush output
Jul 21 07:03:39  | mqttClientAws[7705] | mbedTLS: |2| message length: 48, out_left: 48
Jul 21 07:03:39  | mqttClientAws[7705] | mbedTLS: |2| ssl->f_send() returned 48 (-0xffffffd0)
Jul 21 07:03:39  | mqttClientAws[7705] | mbedTLS: |2| <= flush output
Jul 21 07:03:39  | mqttClientAws[7705] | mbedTLS: |2| <= write record
Jul 21 07:03:39  | mqttClientAws[7705] | mbedTLS: |2| <= write
Jul 21 07:03:39  | mqttClientAws[7705] | [DEBUG] [MQTT] [core_mqtt.c:645] BytesSent=19, BytesRemaining=0^M
Jul 21 07:03:39  | mqttClientAws[7705] | [DEBUG] [MQTT] [core_mqtt.c:666] Successfully sent packet at time 2903398.^M
Jul 21 07:03:39  | mqttClientAws[7705] | [DEBUG] [MQTT] [core_mqtt.c:1809] Sent 19 bytes of CONNECT packet.^M
Jul 21 07:03:39  | mqttClientAws[7705] | mbedTLS: |2| => read
Jul 21 07:03:39  | mqttClientAws[7705] | mbedTLS: |2| => read record
Jul 21 07:03:39  | mqttClientAws[7705] | mbedTLS: |2| => fetch input
Jul 21 07:03:39  | mqttClientAws[7705] | mbedTLS: |2| in_left: 0, nb_want: 5
Jul 21 07:03:49  | mqttClientAws[7705] | mbedTLS: |2| in_left: 0, nb_want: 5
Jul 21 07:03:49  | mqttClientAws[7705] | mbedTLS: |2| ssl->f_recv(_timeout)() returned -26624 (-0x6800)
Jul 21 07:03:49  | mqttClientAws[7705] | mbedTLS: |1| mbedtls_ssl_fetch_input() returned -26624 (-0x6800)
Jul 21 07:03:49  | mqttClientAws[7705] | mbedTLS: |1| ssl_get_next_record() returned -26624 (-0x6800)
Jul 21 07:03:49  | mqttClientAws[7705] | mbedTLS: |1| mbedtls_ssl_read_record() returned -26624 (-0x6800)
Jul 21 07:03:49  | mqttClientAws[7705] | [DEBUG] [Transport_MbedTLS_PKCS11] [mbedtls_pkcs11_posix.c:924] Failed to read data. However, a read can be retried on this error. mbedTLSError= SSL - The operation timed out : <No-Low-Level-Code>.
Jul 21 07:03:49  | mqttClientAws[7705] | [ERROR] [MQTT] [core_mqtt.c:1570] CONNACK recv failed with status = MQTTNoDataAvailable.^M
Jul 21 07:03:49  | mqttClientAws[7705] | [ERROR] [MQTT] [core_mqtt.c:1841] MQTT connection failed with status = MQTTNoDataAvailable.^M
Jul 21 07:03:49  | mqttClientAws[7705] | [ERROR] [MQTT] [mqtt_operations.c:732] Connection with MQTT broker failed with status MQTTNoDataAvailable.^M
Jul 21 07:03:49  | mqttClientAws[7705] | [ERROR] [DEMO] [mqttClientAws.c:1237] Failed to establish MQTT session.^M

--- a/demos/CMakeLists.txt
+++ b/demos/CMakeLists.txt
@@ -9,6 +9,8 @@ else()
     set( PLATFORM_NAME ${PLATFORM_NAME} CACHE STRING "Port to use for building the SDK." )
 endif()
 
+set(AWS_DEMO_INCLUDE_DIRS ${CMAKE_CURRENT_LIST_DIR}/include)
+
 # Include each subdirectory that has a CMakeLists.txt file in it
 file(GLOB demo_dirs "${DEMOS_DIR}/*/*")
 foreach(demo_dir IN LISTS demo_dirs)

--- a/demos/defender/defender_demo_json/CMakeLists.txt
+++ b/demos/defender/defender_demo_json/CMakeLists.txt
@@ -34,6 +34,7 @@ target_include_directories( ${DEMO_NAME} PUBLIC
                             ${BACKOFF_ALGORITHM_INCLUDE_PUBLIC_DIRS}
                             ${JSON_INCLUDE_PUBLIC_DIRS}
                             ${DEFENDER_INCLUDE_PUBLIC_DIRS}
+                            ${AWS_DEMO_INCLUDE_DIRS}
                             ${CMAKE_CURRENT_LIST_DIR} )
 
 set_macro_definitions(TARGETS ${DEMO_NAME}

--- a/demos/defender/defender_demo_json/mqtt_operations.c
+++ b/demos/defender/defender_demo_json/mqtt_operations.c
@@ -56,6 +56,9 @@
 /* Clock for timer. */
 #include "clock.h"
 
+/* AWS IoT Core TLS ALPN definitions for MQTT authentication */
+#include "aws_iot_alpn_defs.h"
+
 /**
  * These configurations are required. Throw compilation error if the below
  * configs are not defined.
@@ -96,25 +99,6 @@
  * @brief Length of the client identifier.
  */
 #define CLIENT_IDENTIFIER_LENGTH                 ( ( uint16_t ) ( sizeof( CLIENT_IDENTIFIER ) - 1 ) )
-
-/**
- * @brief ALPN protocol name for AWS IoT MQTT.
- *
- * This will be used if the AWS_MQTT_PORT is configured as 443 for AWS IoT MQTT
- * broker. Please see more details about the ALPN protocol for AWS IoT MQTT
- * endpoint in the link below.
- * https://aws.amazon.com/blogs/iot/mqtt-with-tls-client-authentication-on-port-443-why-it-is-useful-and-how-it-works/
- *
- * @note OpenSSL requires that the protocol string passed to it for configuration be encoded
- * with the prefix of 8-bit length information of the string. Thus, the 14 byte (0x0e) length
- * information is prefixed to the string.
- */
-#define ALPN_PROTOCOL_NAME                       "\x0ex-amzn-mqtt-ca"
-
-/**
- * @brief Length of ALPN protocol name.
- */
-#define ALPN_PROTOCOL_NAME_LENGTH                ( ( uint16_t ) ( sizeof( ALPN_PROTOCOL_NAME ) - 1 ) )
 
 /**
  * @brief The maximum number of retries for connecting to server.
@@ -378,8 +362,8 @@ static bool connectToBrokerWithBackoffRetries( NetworkContext_t * pNetworkContex
          * in the link below.
          * https://aws.amazon.com/blogs/iot/mqtt-with-tls-client-authentication-on-port-443-why-it-is-useful-and-how-it-works/
          */
-        opensslCredentials.pAlpnProtos = ALPN_PROTOCOL_NAME;
-        opensslCredentials.alpnProtosLen = ALPN_PROTOCOL_NAME_LENGTH;
+        opensslCredentials.pAlpnProtos = AWS_IOT_ALPN_MQTT_CA_AUTH_OPENSSL;
+        opensslCredentials.alpnProtosLen = AWS_IOT_ALPN_MQTT_CA_AUTH_OPENSSL_LEN;
     }
 
     /* Seed pseudo random number generator used in the demo for

--- a/demos/defender/defender_demo_json/mqtt_operations.c
+++ b/demos/defender/defender_demo_json/mqtt_operations.c
@@ -56,7 +56,7 @@
 /* Clock for timer. */
 #include "clock.h"
 
-/* AWS IoT Core TLS ALPN definitions for MQTT authentication */
+/* AWS IoT Core TLS ALPN definitions for MQTT authentication. */
 #include "aws_iot_alpn_defs.h"
 
 /**

--- a/demos/fleet_provisioning/fleet_provisioning_with_csr/CMakeLists.txt
+++ b/demos/fleet_provisioning/fleet_provisioning_with_csr/CMakeLists.txt
@@ -47,6 +47,7 @@ target_include_directories( ${DEMO_NAME}
                               ${BACKOFF_ALGORITHM_INCLUDE_PUBLIC_DIRS}
                               ${PKCS_INCLUDE_PUBLIC_DIRS}
                               ${PKCS_PAL_INCLUDE_PUBLIC_DIRS}
+                              ${AWS_DEMO_INCLUDE_DIRS}
                               "${FLEET_PROVISIONING_INCLUDE_PUBLIC_DIRS}"
                               "${DEMOS_DIR}/pkcs11/common/include" # corePKCS11 config
                               "${CMAKE_SOURCE_DIR}/platform/include"

--- a/demos/fleet_provisioning/fleet_provisioning_with_csr/mqtt_operations.c
+++ b/demos/fleet_provisioning/fleet_provisioning_with_csr/mqtt_operations.c
@@ -56,6 +56,9 @@
 /* Clock for timer. */
 #include "clock.h"
 
+/* AWS IoT Core TLS ALPN definitions for MQTT authentication */
+#include "aws_iot_alpn_defs.h"
+
 /**
  * These configurations are required. Throw compilation error if the below
  * configs are not defined.
@@ -90,25 +93,6 @@
  * @brief Length of the client identifier.
  */
 #define CLIENT_IDENTIFIER_LENGTH                 ( ( uint16_t ) ( sizeof( CLIENT_IDENTIFIER ) - 1 ) )
-
-/**
- * @brief ALPN protocol name for AWS IoT MQTT.
- *
- * This will be used if the AWS_MQTT_PORT is configured as 443 for AWS IoT MQTT
- * broker. Please see more details about the ALPN protocol for AWS IoT MQTT
- * endpoint in the link below.
- * https://aws.amazon.com/blogs/iot/mqtt-with-tls-client-authentication-on-port-443-why-it-is-useful-and-how-it-works/
- *
- * @note OpenSSL requires that the protocol string passed to it for configuration be encoded
- * with the prefix of 8-bit length information of the string. Thus, the 14 byte (0x0e) length
- * information is prefixed to the string.
- */
-#define ALPN_PROTOCOL_NAME                       "\x0ex-amzn-mqtt-ca"
-
-/**
- * @brief Length of ALPN protocol name.
- */
-#define ALPN_PROTOCOL_NAME_LENGTH                ( ( uint16_t ) ( sizeof( ALPN_PROTOCOL_NAME ) - 1 ) )
 
 /**
  * @brief The maximum number of retries for connecting to server.
@@ -356,7 +340,6 @@ static bool connectToBrokerWithBackoffRetries( NetworkContext_t * pNetworkContex
     BackoffAlgorithmContext_t reconnectParams;
     MbedtlsPkcs11Credentials_t tlsCredentials = { 0 };
     uint16_t nextRetryBackOff = 0U;
-    const char * alpn[] = { ALPN_PROTOCOL_NAME, NULL };
 
     /* Set the pParams member of the network context with desired transport. */
     pNetworkContext->pParams = &tlsContext;
@@ -377,12 +360,14 @@ static bool connectToBrokerWithBackoffRetries( NetworkContext_t * pNetworkContex
 
     if( AWS_MQTT_PORT == 443 )
     {
+        static const char * alpnProtoArray[] = AWS_IOT_ALPN_MQTT_CA_AUTH_MBEDTLS;
+
         /* Pass the ALPN protocol name depending on the port being used.
          * Please see more details about the ALPN protocol for AWS IoT MQTT endpoint
          * in the link below.
          * https://aws.amazon.com/blogs/iot/mqtt-with-tls-client-authentication-on-port-443-why-it-is-useful-and-how-it-works/
          */
-        tlsCredentials.pAlpnProtos = alpn;
+        tlsCredentials.pAlpnProtos = alpnProtoArray;
     }
 
     /* Initialize reconnect attempts and interval */

--- a/demos/http/http_demo_mutual_auth/CMakeLists.txt
+++ b/demos/http/http_demo_mutual_auth/CMakeLists.txt
@@ -40,8 +40,9 @@ target_include_directories(
         "${DEMOS_DIR}/http/common/include"
         ${HTTP_INCLUDE_PUBLIC_DIRS}
         ${BACKOFF_ALGORITHM_INCLUDE_PUBLIC_DIRS}
-        ${CMAKE_CURRENT_LIST_DIR}
         ${LOGGING_INCLUDE_DIRS}
+        ${AWS_DEMO_INCLUDE_DIRS}
+        ${CMAKE_CURRENT_LIST_DIR}
 )
 
 set_macro_definitions(TARGETS ${DEMO_NAME}

--- a/demos/http/http_demo_mutual_auth/http_demo_mutual_auth.c
+++ b/demos/http/http_demo_mutual_auth/http_demo_mutual_auth.c
@@ -40,6 +40,9 @@
 /* OpenSSL transport header. */
 #include "openssl_posix.h"
 
+/* AWS IoT Core TLS ALPN definitions for MQTT authentication */
+#include "aws_iot_alpn_defs.h"
+
 /* Check that AWS IoT Core endpoint is defined. */
 #ifndef AWS_IOT_ENDPOINT
     #error "AWS_IOT_ENDPOINT must be defined to your AWS IoT Core endpoint."
@@ -69,17 +72,6 @@
 #ifndef CLIENT_PRIVATE_KEY_PATH
     #error "Please define a CLIENT_PRIVATE_KEY_PATH."
 #endif
-
-/**
- * @brief ALPN protocol name to be sent as part of the ClientHello message.
- *
- * @note When using ALPN, port 443 must be used to connect to AWS IoT Core.
- *
- * @note OpenSSL requires that the protocol string passed to it for configuration be encoded
- * with the prefix of 8-bit length information of the string. Thus, the 14 byte (0x0e)
- * length information is prefixed to the string.
- */
-#define IOT_CORE_ALPN_PROTOCOL_NAME    "\x0ex-amzn-http-ca"
 
 /* Check that transport timeout for transport send and receive is defined. */
 #ifndef TRANSPORT_SEND_RECV_TIMEOUT_MS
@@ -185,8 +177,8 @@ static int32_t connectToServer( NetworkContext_t * pNetworkContext )
     /* ALPN is required when communicating to AWS IoT Core over port 443 through HTTP. */
     if( AWS_HTTPS_PORT == 443 )
     {
-        opensslCredentials.pAlpnProtos = IOT_CORE_ALPN_PROTOCOL_NAME;
-        opensslCredentials.alpnProtosLen = strlen( IOT_CORE_ALPN_PROTOCOL_NAME );
+        opensslCredentials.pAlpnProtos = AWS_IOT_ALPN_HTTP_CA_AUTH_OPENSSL;
+        opensslCredentials.alpnProtosLen = AWS_IOT_ALPN_MQTT_CA_AUTH_OPENSSL_LEN;
     }
 
     /* Initialize server information. */

--- a/demos/include/aws_iot_alpn_defs.h
+++ b/demos/include/aws_iot_alpn_defs.h
@@ -1,0 +1,65 @@
+/*
+ * AWS IoT Device SDK for Embedded C 202108.00
+ * Copyright (C) 2020 Amazon.com, Inc. or its affiliates.  All Rights Reserved.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy of
+ * this software and associated documentation files (the "Software"), to deal in
+ * the Software without restriction, including without limitation the rights to
+ * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
+ * the Software, and to permit persons to whom the Software is furnished to do so,
+ * subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+#ifndef AWS_IOT_ALPN_DEFS_H_
+#define AWS_IOT_ALPN_DEFS_H_
+
+/*
+ * AWS IoT Core allows the use of HTTP or MQTT over TCP port 443. Connections
+ * to IoT Core endpoints on port 443 default to the HTTP protocol.
+ *
+ * Clients must use the TLS Application-Layer Protocol Negotiation extension
+ * to notify the service of it's desired protocol or authentication mode.
+ *
+ * For more information, please reference the following URL:
+ * https://docs.aws.amazon.com/iot/latest/developerguide/protocols.html
+ */
+
+#define AWS_IOT_ALPN_MQTT_CUSTOM_AUTH                "mqtt"
+#define AWS_IOT_ALPN_MQTT_CA_AUTH                    "x-amzn-mqtt-ca"
+#define AWS_IOT_ALPN_HTTP_CA_AUTH                    "x-amzn-http-ca"
+
+/*
+ * @note OpenSSL requires that ALPN identifiers be provided in protocol form.
+ *       This means that each alpn string be prepended with its length.
+ *       When multiple concurrent alpn identifiers are necessary,
+ *       each identifier should be prepended with it's respective length and
+ *       then concatenated together.
+ */
+#define AWS_IOT_ALPN_MQTT_CUSTOM_AUTH_OPENSSL        "\x04mqtt"
+#define AWS_IOT_ALPN_MQTT_CUSTOM_AUTH_OPENSSL_LEN    ( sizeof( AWS_IOT_ALPN_MQTT_CUSTOM_AUTH_OPENSSL ) - 1U )
+
+#define AWS_IOT_ALPN_MQTT_CA_AUTH_OPENSSL            "\x0ex-amzn-mqtt-ca"
+#define AWS_IOT_ALPN_MQTT_CA_AUTH_OPENSSL_LEN        ( sizeof( AWS_IOT_ALPN_MQTT_CA_AUTH_OPENSSL ) - 1U )
+
+#define AWS_IOT_ALPN_HTTP_CA_AUTH_OPENSSL            "\x0ex-amzn-http-ca"
+#define AWS_IOT_ALPN_HTTP_CA_AUTH_OPENSSL_LEN        ( sizeof( AWS_IOT_ALPN_HTTP_CA_AUTH_OPENSSL ) - 1U )
+
+/*
+ * @note MbedTLS requires alpn identifiers to be provided in a null terminated
+ *       array of null terminated c strings.
+ */
+#define AWS_IOT_ALPN_MQTT_CUSTOM_AUTH_MBEDTLS        { AWS_IOT_ALPN_MQTT_CUSTOM_AUTH, NULL }
+#define AWS_IOT_ALPN_MQTT_CA_AUTH_MBEDTLS            { AWS_IOT_ALPN_MQTT_CA_AUTH, NULL }
+#define AWS_IOT_ALPN_HTTP_CA_AUTH_MBEDTLS            { AWS_IOT_ALPN_HTTP_CA_AUTH, NULL }
+
+#endif /* AWS_IOT_ALPN_DEFS_H_ */

--- a/demos/mqtt/mqtt_demo_mutual_auth/CMakeLists.txt
+++ b/demos/mqtt/mqtt_demo_mutual_auth/CMakeLists.txt
@@ -30,8 +30,9 @@ target_include_directories(
     PUBLIC
         ${MQTT_INCLUDE_PUBLIC_DIRS}
         ${BACKOFF_ALGORITHM_INCLUDE_PUBLIC_DIRS}
-        ${CMAKE_CURRENT_LIST_DIR}
         ${LOGGING_INCLUDE_DIRS}
+        ${AWS_DEMO_INCLUDE_DIRS}
+        ${CMAKE_CURRENT_LIST_DIR}
 )
 
 set_macro_definitions(TARGETS ${DEMO_NAME}

--- a/demos/mqtt/mqtt_demo_mutual_auth/mqtt_demo_mutual_auth.c
+++ b/demos/mqtt/mqtt_demo_mutual_auth/mqtt_demo_mutual_auth.c
@@ -75,6 +75,9 @@
 /* Clock for timer. */
 #include "clock.h"
 
+/* AWS IoT Core TLS ALPN definitions for MQTT authentication */
+#include "aws_iot_alpn_defs.h"
+
 /**
  * These configuration settings are required to run the mutual auth demo.
  * Throw compilation error if the below configs are not defined.
@@ -139,47 +142,12 @@
 /**
  * @brief Length of MQTT server host name.
  */
-#define AWS_IOT_ENDPOINT_LENGTH         ( ( uint16_t ) ( sizeof( AWS_IOT_ENDPOINT ) - 1 ) )
+#define AWS_IOT_ENDPOINT_LENGTH                  ( ( uint16_t ) ( sizeof( AWS_IOT_ENDPOINT ) - 1 ) )
 
 /**
  * @brief Length of client identifier.
  */
-#define CLIENT_IDENTIFIER_LENGTH        ( ( uint16_t ) ( sizeof( CLIENT_IDENTIFIER ) - 1 ) )
-
-/**
- * @brief ALPN (Application-Layer Protocol Negotiation) protocol name for AWS IoT MQTT.
- *
- * This will be used if the AWS_MQTT_PORT is configured as 443 for AWS IoT MQTT broker.
- * Please see more details about the ALPN protocol for AWS IoT MQTT endpoint
- * in the link below.
- * https://aws.amazon.com/blogs/iot/mqtt-with-tls-client-authentication-on-port-443-why-it-is-useful-and-how-it-works/
- *
- * @note OpenSSL requires that the protocol string passed to it for configuration be encoded
- * with the prefix of 8-bit length information of the string. Thus, the 14 byte (0x0e) length
- * information is prefixed to the string.
- */
-#define AWS_IOT_MQTT_ALPN               "\x0ex-amzn-mqtt-ca"
-
-/**
- * @brief Length of ALPN protocol name.
- */
-#define AWS_IOT_MQTT_ALPN_LENGTH        ( ( uint16_t ) ( sizeof( AWS_IOT_MQTT_ALPN ) - 1 ) )
-
-/**
- * @brief This is the ALPN (Application-Layer Protocol Negotiation) string
- * required by AWS IoT for password-based authentication using TCP port 443.
- *
- * @note OpenSSL requires that the protocol string passed to it for configuration
- * be encoded with the prefix of 8-bit length information of the string. Thus, the
- * 4 byte (0x04) length information is prefixed to the string.
- */
-#define AWS_IOT_PASSWORD_ALPN           "\x04mqtt"
-
-/**
- * @brief Length of password ALPN.
- */
-#define AWS_IOT_PASSWORD_ALPN_LENGTH    ( ( uint16_t ) ( sizeof( AWS_IOT_PASSWORD_ALPN ) - 1 ) )
-
+#define CLIENT_IDENTIFIER_LENGTH                 ( ( uint16_t ) ( sizeof( CLIENT_IDENTIFIER ) - 1 ) )
 
 /**
  * @brief The maximum number of retries for connecting to server.
@@ -201,46 +169,45 @@
  */
 #define CONNACK_RECV_TIMEOUT_MS                  ( 1000U )
 
-
 /**
  * @brief The topic to subscribe and publish to in the example.
  *
  * The topic name starts with the client identifier to ensure that each demo
  * interacts with a unique topic name.
  */
-#define MQTT_EXAMPLE_TOPIC                  CLIENT_IDENTIFIER "/example/topic"
+#define MQTT_EXAMPLE_TOPIC                       CLIENT_IDENTIFIER "/example/topic"
 
 /**
  * @brief Length of client MQTT topic.
  */
-#define MQTT_EXAMPLE_TOPIC_LENGTH           ( ( uint16_t ) ( sizeof( MQTT_EXAMPLE_TOPIC ) - 1 ) )
+#define MQTT_EXAMPLE_TOPIC_LENGTH                ( ( uint16_t ) ( sizeof( MQTT_EXAMPLE_TOPIC ) - 1 ) )
 
 /**
  * @brief The MQTT message published in this example.
  */
-#define MQTT_EXAMPLE_MESSAGE                "Hello World!"
+#define MQTT_EXAMPLE_MESSAGE                     "Hello World!"
 
 /**
  * @brief The length of the MQTT message published in this example.
  */
-#define MQTT_EXAMPLE_MESSAGE_LENGTH         ( ( uint16_t ) ( sizeof( MQTT_EXAMPLE_MESSAGE ) - 1 ) )
+#define MQTT_EXAMPLE_MESSAGE_LENGTH              ( ( uint16_t ) ( sizeof( MQTT_EXAMPLE_MESSAGE ) - 1 ) )
 
 /**
  * @brief Maximum number of outgoing publishes maintained in the application
  * until an ack is received from the broker.
  */
-#define MAX_OUTGOING_PUBLISHES              ( 5U )
+#define MAX_OUTGOING_PUBLISHES                   ( 5U )
 
 /**
  * @brief Invalid packet identifier for the MQTT packets. Zero is always an
  * invalid packet identifier as per MQTT 3.1.1 spec.
  */
-#define MQTT_PACKET_ID_INVALID              ( ( uint16_t ) 0U )
+#define MQTT_PACKET_ID_INVALID                   ( ( uint16_t ) 0U )
 
 /**
  * @brief Timeout for MQTT_ProcessLoop function in milliseconds.
  */
-#define MQTT_PROCESS_LOOP_TIMEOUT_MS        ( 500U )
+#define MQTT_PROCESS_LOOP_TIMEOUT_MS             ( 500U )
 
 /**
  * @brief The maximum time interval in seconds which is allowed to elapse
@@ -251,37 +218,37 @@
  *  absence of sending any other Control Packets, the Client MUST send a
  *  PINGREQ Packet.
  */
-#define MQTT_KEEP_ALIVE_INTERVAL_SECONDS    ( 60U )
+#define MQTT_KEEP_ALIVE_INTERVAL_SECONDS         ( 60U )
 
 /**
  * @brief Delay between MQTT publishes in seconds.
  */
-#define DELAY_BETWEEN_PUBLISHES_SECONDS     ( 1U )
+#define DELAY_BETWEEN_PUBLISHES_SECONDS          ( 1U )
 
 /**
  * @brief Number of PUBLISH messages sent per iteration.
  */
-#define MQTT_PUBLISH_COUNT_PER_LOOP         ( 5U )
+#define MQTT_PUBLISH_COUNT_PER_LOOP              ( 5U )
 
 /**
  * @brief Delay in seconds between two iterations of subscribePublishLoop().
  */
-#define MQTT_SUBPUB_LOOP_DELAY_SECONDS      ( 5U )
+#define MQTT_SUBPUB_LOOP_DELAY_SECONDS           ( 5U )
 
 /**
  * @brief Transport timeout in milliseconds for transport send and receive.
  */
-#define TRANSPORT_SEND_RECV_TIMEOUT_MS      ( 500 )
+#define TRANSPORT_SEND_RECV_TIMEOUT_MS           ( 500 )
 
 /**
  * @brief The MQTT metrics string expected by AWS IoT.
  */
-#define METRICS_STRING                      "?SDK=" OS_NAME "&Version=" OS_VERSION "&Platform=" HARDWARE_PLATFORM_NAME "&MQTTLib=" MQTT_LIB
+#define METRICS_STRING                           "?SDK=" OS_NAME "&Version=" OS_VERSION "&Platform=" HARDWARE_PLATFORM_NAME "&MQTTLib=" MQTT_LIB
 
 /**
  * @brief The length of the MQTT metrics string expected by AWS IoT.
  */
-#define METRICS_STRING_LENGTH               ( ( uint16_t ) ( sizeof( METRICS_STRING ) - 1 ) )
+#define METRICS_STRING_LENGTH                    ( ( uint16_t ) ( sizeof( METRICS_STRING ) - 1 ) )
 
 
 #ifdef CLIENT_USERNAME
@@ -605,22 +572,22 @@ static int connectToServerWithBackoffRetries( NetworkContext_t * pNetworkContext
 
     if( AWS_MQTT_PORT == 443 )
     {
-        /* Pass the ALPN protocol name depending on the port being used.
+        /* Pass the ALPN protocol name depending on the port and auth type being used.
          * Please see more details about the ALPN protocol for the AWS IoT MQTT
          * endpoint in the link below.
          * https://aws.amazon.com/blogs/iot/mqtt-with-tls-client-authentication-on-port-443-why-it-is-useful-and-how-it-works/
          *
          * For username and password based authentication in AWS IoT,
-         * #AWS_IOT_PASSWORD_ALPN is used. More details can be found in the
+         * #AWS_IOT_ALPN_MQTT_CUSTOM_AUTH is used. More details can be found in the
          * link below.
          * https://docs.aws.amazon.com/iot/latest/developerguide/custom-authentication.html
          */
         #ifdef CLIENT_USERNAME
-            opensslCredentials.pAlpnProtos = AWS_IOT_PASSWORD_ALPN;
-            opensslCredentials.alpnProtosLen = AWS_IOT_PASSWORD_ALPN_LENGTH;
+            opensslCredentials.pAlpnProtos = AWS_IOT_ALPN_MQTT_CUSTOM_AUTH_OPENSSL;
+            opensslCredentials.alpnProtosLen = AWS_IOT_ALPN_MQTT_CUSTOM_AUTH_OPENSSL_LEN;
         #else
-            opensslCredentials.pAlpnProtos = AWS_IOT_MQTT_ALPN;
-            opensslCredentials.alpnProtosLen = AWS_IOT_MQTT_ALPN_LENGTH;
+            opensslCredentials.pAlpnProtos = AWS_IOT_ALPN_MQTT_CA_AUTH_OPENSSL;
+            opensslCredentials.alpnProtosLen = AWS_IOT_ALPN_MQTT_CA_AUTH_OPENSSL_LEN;
         #endif
     }
 

--- a/demos/ota/ota_demo_core_http/CMakeLists.txt
+++ b/demos/ota/ota_demo_core_http/CMakeLists.txt
@@ -54,6 +54,7 @@ target_include_directories(
         ${OTA_INCLUDE_OS_POSIX_DIRS}
         ${MQTT_INCLUDE_PUBLIC_DIRS}
         ${HTTP_INCLUDE_PUBLIC_DIRS}
+        ${AWS_DEMO_INCLUDE_DIRS}
 )
 
 set_macro_definitions(TARGETS ${DEMO_NAME}

--- a/demos/ota/ota_demo_core_http/ota_demo_core_http.c
+++ b/demos/ota/ota_demo_core_http/ota_demo_core_http.c
@@ -69,6 +69,9 @@
 /* Include firmware version struct definition. */
 #include "ota_appversion32.h"
 
+/* AWS IoT Core TLS ALPN definitions for MQTT authentication */
+#include "aws_iot_alpn_defs.h"
+
 /**
  * These configuration settings are required to run the OTA demo which uses mutual authentication.
  * Throw compilation error if the below configs are not defined.
@@ -91,25 +94,6 @@
 #ifndef CLIENT_PRIVATE_KEY_PATH
     #error "Please define path to client private key(CLIENT_PRIVATE_KEY_PATH) in demo_config.h."
 #endif
-
-/**
- * @brief ALPN (Application-Layer Protocol Negotiation) protocol name for AWS IoT MQTT.
- *
- * This will be used if the AWS_MQTT_PORT is configured as 443 for AWS IoT MQTT broker.
- * Please see more details about the ALPN protocol for AWS IoT MQTT endpoint
- * in the link below.
- * https://aws.amazon.com/blogs/iot/mqtt-with-tls-client-authentication-on-port-443-why-it-is-useful-and-how-it-works/
- *
- * @note OpenSSL requires that the protocol string passed to it for configuration be encoded
- * with the prefix of 8-bit length information of the string. Thus, the 14 byte (0x0e) length
- * information is prefixed to the string.
- */
-#define AWS_IOT_MQTT_ALPN                        "\x0ex-amzn-mqtt-ca"
-
-/**
- * @brief Length of ALPN protocol name.
- */
-#define AWS_IOT_MQTT_ALPN_LENGTH                 ( ( uint16_t ) ( sizeof( AWS_IOT_MQTT_ALPN ) - 1 ) )
 
 /**
  * @brief Length of MQTT server host name.
@@ -1089,11 +1073,11 @@ static int connectToServerWithBackoffRetries( NetworkContext_t * pNetworkContext
          * https://docs.aws.amazon.com/iot/latest/developerguide/custom-authentication.html
          */
         #ifdef CLIENT_USERNAME
-            opensslCredentials.pAlpnProtos = AWS_IOT_PASSWORD_ALPN;
-            opensslCredentials.alpnProtosLen = AWS_IOT_PASSWORD_ALPN_LENGTH;
+            opensslCredentials.pAlpnProtos = AWS_IOT_ALPN_MQTT_CUSTOM_AUTH_OPENSSL;
+            opensslCredentials.alpnProtosLen = AWS_IOT_ALPN_MQTT_CUSTOM_AUTH_OPENSSL_LEN;
         #else
-            opensslCredentials.pAlpnProtos = AWS_IOT_MQTT_ALPN;
-            opensslCredentials.alpnProtosLen = AWS_IOT_MQTT_ALPN_LENGTH;
+            opensslCredentials.pAlpnProtos = AWS_IOT_ALPN_MQTT_CA_AUTH_OPENSSL;
+            opensslCredentials.alpnProtosLen = AWS_IOT_ALPN_MQTT_CA_AUTH_OPENSSL_LEN;
         #endif
     }
 

--- a/demos/ota/ota_demo_core_mqtt/CMakeLists.txt
+++ b/demos/ota/ota_demo_core_mqtt/CMakeLists.txt
@@ -41,6 +41,7 @@ target_include_directories(
         ${OTA_INCLUDE_PRIVATE_DIRS}
         ${OTA_INCLUDE_OS_POSIX_DIRS}
         ${MQTT_INCLUDE_PUBLIC_DIRS}
+        ${AWS_DEMO_INCLUDE_DIRS}
 )
 
 set_macro_definitions(TARGETS ${DEMO_NAME}

--- a/demos/ota/ota_demo_core_mqtt/ota_demo_core_mqtt.c
+++ b/demos/ota/ota_demo_core_mqtt/ota_demo_core_mqtt.c
@@ -63,6 +63,9 @@
 /* Include firmware version struct definition. */
 #include "ota_appversion32.h"
 
+/* AWS IoT Core TLS ALPN definitions for MQTT authentication */
+#include "aws_iot_alpn_defs.h"
+
 /**
  * These configuration settings are required to run the OTA demo which uses mutual authentication.
  * Throw compilation error if the below configs are not defined.
@@ -82,25 +85,6 @@
 #ifndef CLIENT_PRIVATE_KEY_PATH
     #error "Please define path to client private key(CLIENT_PRIVATE_KEY_PATH) in demo_config.h."
 #endif
-
-/**
- * @brief ALPN (Application-Layer Protocol Negotiation) protocol name for AWS IoT MQTT.
- *
- * This will be used if the AWS_MQTT_PORT is configured as 443 for AWS IoT MQTT broker.
- * Please see more details about the ALPN protocol for AWS IoT MQTT endpoint
- * in the link below.
- * https://aws.amazon.com/blogs/iot/mqtt-with-tls-client-authentication-on-port-443-why-it-is-useful-and-how-it-works/
- *
- * @note OpenSSL requires that the protocol string passed to it for configuration be encoded
- * with the prefix of 8-bit length information of the string. Thus, the 14 byte (0x0e) length
- * information is prefixed to the string.
- */
-#define AWS_IOT_MQTT_ALPN                   "\x0ex-amzn-mqtt-ca"
-
-/**
- * @brief Length of ALPN protocol name.
- */
-#define AWS_IOT_MQTT_ALPN_LENGTH            ( ( uint16_t ) ( sizeof( AWS_IOT_MQTT_ALPN ) - 1 ) )
 
 /**
  * @brief Length of MQTT server host name.
@@ -958,11 +942,11 @@ static int connectToServerWithBackoffRetries( NetworkContext_t * pNetworkContext
          * https://docs.aws.amazon.com/iot/latest/developerguide/custom-authentication.html
          */
         #ifdef CLIENT_USERNAME
-            opensslCredentials.pAlpnProtos = AWS_IOT_PASSWORD_ALPN;
-            opensslCredentials.alpnProtosLen = AWS_IOT_PASSWORD_ALPN_LENGTH;
+            opensslCredentials.pAlpnProtos = AWS_IOT_ALPN_MQTT_CUSTOM_AUTH_OPENSSL;
+            opensslCredentials.alpnProtosLen = AWS_IOT_ALPN_MQTT_CUSTOM_AUTH_OPENSSL_LEN;
         #else
-            opensslCredentials.pAlpnProtos = AWS_IOT_MQTT_ALPN;
-            opensslCredentials.alpnProtosLen = AWS_IOT_MQTT_ALPN_LENGTH;
+            opensslCredentials.pAlpnProtos = AWS_IOT_ALPN_MQTT_CA_AUTH_OPENSSL;
+            opensslCredentials.alpnProtosLen = AWS_IOT_ALPN_MQTT_CA_AUTH_OPENSSL_LEN;
         #endif
     }
 

--- a/demos/shadow/shadow_demo_main/CMakeLists.txt
+++ b/demos/shadow/shadow_demo_main/CMakeLists.txt
@@ -40,8 +40,9 @@ target_include_directories(
         ${MQTT_INCLUDE_PUBLIC_DIRS}
         ${BACKOFF_ALGORITHM_INCLUDE_PUBLIC_DIRS}
         ${SHADOW_INCLUDE_PUBLIC_DIRS}
-        ${CMAKE_CURRENT_LIST_DIR}
         ${JSON_INCLUDE_PUBLIC_DIRS}
+        ${AWS_DEMO_INCLUDE_DIRS}
+        ${CMAKE_CURRENT_LIST_DIR}
 )
 
 set_macro_definitions(TARGETS ${DEMO_NAME}

--- a/demos/shadow/shadow_demo_main/shadow_demo_helpers.c
+++ b/demos/shadow/shadow_demo_main/shadow_demo_helpers.c
@@ -53,6 +53,8 @@
 /* Clock for timer. */
 #include "clock.h"
 
+/* AWS IoT Core TLS ALPN definitions for MQTT authentication */
+#include "aws_iot_alpn_defs.h"
 
 /**
  * These configuration settings are required to run the shadow demo.
@@ -88,32 +90,12 @@
 /**
  * @brief Length of MQTT server host name.
  */
-#define AWS_IOT_ENDPOINT_LENGTH      ( ( uint16_t ) ( sizeof( AWS_IOT_ENDPOINT ) - 1 ) )
+#define AWS_IOT_ENDPOINT_LENGTH                  ( ( uint16_t ) ( sizeof( AWS_IOT_ENDPOINT ) - 1 ) )
 
 /**
  * @brief Length of client identifier.
  */
-#define CLIENT_IDENTIFIER_LENGTH     ( ( uint16_t ) ( sizeof( CLIENT_IDENTIFIER ) - 1 ) )
-
-/**
- * @brief ALPN protocol name for AWS IoT MQTT.
- *
- * This will be used if the AWS_MQTT_PORT is configured as 443 for AWS IoT MQTT broker.
- * Please see more details about the ALPN protocol for AWS IoT MQTT endpoint
- * in the link below.
- * https://aws.amazon.com/blogs/iot/mqtt-with-tls-client-authentication-on-port-443-why-it-is-useful-and-how-it-works/
- *
- * @note OpenSSL requires that the protocol string passed to it for configuration be encoded
- * with the prefix of 8-bit length information of the string. Thus, the 14 byte (0x0e) length
- * information is prefixed to the string.
- */
-#define ALPN_PROTOCOL_NAME           "\x0ex-amzn-mqtt-ca"
-
-/**
- * @brief Length of ALPN protocol name.
- */
-#define ALPN_PROTOCOL_NAME_LENGTH    ( ( uint16_t ) ( sizeof( ALPN_PROTOCOL_NAME ) - 1 ) )
-
+#define CLIENT_IDENTIFIER_LENGTH                 ( ( uint16_t ) ( sizeof( CLIENT_IDENTIFIER ) - 1 ) )
 
 /**
  * @brief The maximum number of retries for connecting to server.
@@ -361,8 +343,8 @@ static int connectToServerWithBackoffRetries( NetworkContext_t * pNetworkContext
          * in the link below.
          * https://aws.amazon.com/blogs/iot/mqtt-with-tls-client-authentication-on-port-443-why-it-is-useful-and-how-it-works/
          */
-        opensslCredentials.pAlpnProtos = ALPN_PROTOCOL_NAME;
-        opensslCredentials.alpnProtosLen = ALPN_PROTOCOL_NAME_LENGTH;
+        opensslCredentials.pAlpnProtos = AWS_IOT_ALPN_MQTT_CA_AUTH_OPENSSL;
+        opensslCredentials.alpnProtosLen = AWS_IOT_ALPN_MQTT_CA_AUTH_OPENSSL_LEN;
     }
 
     /* Seed pseudo random number generator used in the demo for

